# Проект по SQL

## Содержание проекта

**Описание проекта**

**Цель проекта**

**Описание данных**

**Схема данных**

**Задачи**

**Общая информация о данных**

**Анализ данных**

**Общий вывод**

**Рекомендации**


## Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

## Цель проекта

Чтобы получить целостное представление о продукте, необходимо изучить базы данных и найти в них целевую информацию.

## Описание данных

**Таблица books**

Содержит данные о книгах:

book_id — идентификатор книги;

author_id — идентификатор автора;

title — название книги;

num_pages — количество страниц;

publication_date — дата публикации книги;

publisher_id — идентификатор издателя.

**Таблица authors**

Содержит данные об авторах:

author_id — идентификатор автора;

author — имя автора.

Таблица publishers

Содержит данные об издательствах:

publisher_id — идентификатор издательства;

publisher — название издательства;

**Таблица ratings**

Содержит данные о пользовательских оценках книг:

rating_id — идентификатор оценки;

book_id — идентификатор книги;

username — имя пользователя, оставившего оценку;

rating — оценка книги.

**Таблица reviews**

Содержит данные о пользовательских обзорах:

review_id — идентификатор обзора;

book_id — идентификатор книги;

username — имя автора обзора;

text — текст обзора.

## Анализ данных

- Посчитаем, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитаем количество обзоров и среднюю оценку;
- Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры;
- Определим автора с самой высокой средней оценкой книг — учитываем только книги с 50 и более оценками;
- Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

# Общая информация о данных

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [2]:
# функция для выполнения SQL-запросов и возврата результатов в виде DataFrame
def execute_sql_query(query):
    connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(
        db_config['user'],
        db_config['pwd'],
        db_config['host'],
        db_config['port'],
        db_config['db']
    )
    # Создаем коннектор
    engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 
    # Выполняем SQL-запрос и сохраняем результаты в DataFrame
    df = pd.read_sql(query, con=engine)
    # Возвращаем DataFrame с результатами
    return df


In [3]:
# Чтение данных из таблицы books
query = 'SELECT * FROM books'
df_books = execute_sql_query(query)
df_books

book_id  author_id                                              title  \
0          1        546                                       'Salem's Lot   
1          2        465                 1 000 Places to See Before You Die   
2          3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3          4         82  1491: New Revelations of the Americas Before C...   
4          5        125                                               1776   
..       ...        ...                                                ...   
995      996        571           Wyrd Sisters (Discworld  #6; Witches #2)   
996      997        454                        Xenocide (Ender's Saga  #3)   
997      998        201                                    Year of Wonders   
998      999         94                        You Suck (A Love Story  #2)   
999     1000        509  Zen and the Art of Motorcycle Maintenance: An ...   

     num_pages publication_date  publisher_id  
0          594       2005-11-01            93  
1          992       2003-05-22           336  
2          322       2010-12-21           135  
3          541       2006-10-10           309  
4          386       2006-07-04           268  
..         ...              ...           ...  
995        265       2001-02-06           147  
996        592       1996-07-15           297  
997        358       2002-04-30           212  
998        328       2007-01-16           331  
999        540       2006-04-25           143  

[1000 rows x 6 columns]

In [4]:
# Чтение данных из таблицы authors
query = 'SELECT * FROM authors'
df_authors = execute_sql_query(query)
df_authors

author_id                                author
0            1                            A.S. Byatt
1            2        Aesop/Laura Harris/Laura Gibbs
2            3                       Agatha Christie
3            4                         Alan Brennert
4            5              Alan Moore/David   Lloyd
..         ...                                   ...
631        632         William Strunk Jr./E.B. White
632        633                           Zadie Smith
633        634                 Zilpha Keatley Snyder
634        635                    Zora Neale Hurston
635        636  Åsne Seierstad/Ingrid Christopherson

[636 rows x 2 columns]

In [5]:
# Чтение данных из таблицы publishers
query = 'SELECT * FROM publishers'
df_publishers = execute_sql_query(query)
df_publishers

publisher_id                          publisher
0               1                                Ace
1               2                           Ace Book
2               3                          Ace Books
3               4                      Ace Hardcover
4               5  Addison Wesley Publishing Company
..            ...                                ...
335           336         Workman Publishing Company
336           337                         Wyatt Book
337           338              Yale University Press
338           339                           Yearling
339           340                     Yearling Books

[340 rows x 2 columns]

In [6]:
# Чтение данных из таблицы ratings
query = 'SELECT * FROM ratings'
df_ratings = execute_sql_query(query)
df_ratings

rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

In [7]:
# Чтение данных из таблицы reviews
query = 'SELECT * FROM reviews'
df_reviews = execute_sql_query(query)
df_ratings


rating_id  book_id        username  rating
0             1        1      ryanfranco       4
1             2        1   grantpatricia       2
2             3        1    brandtandrea       5
3             4        2        lorichen       3
4             5        2     mariokeller       2
...         ...      ...             ...     ...
6451       6452     1000  carolrodriguez       4
6452       6453     1000         wendy18       4
6453       6454     1000      jarvispaul       5
6454       6455     1000           zross       2
6455       6456     1000         fharris       5

[6456 rows x 4 columns]

## Решение задач

### Посчитаем, сколько книг вышло после 1 января 2000 года

In [8]:
# SQL-запрос
query = "SELECT COUNT(book_id) AS book_count FROM books WHERE publication_date > '2000-01-01';"

# Выполнение SQL-запроса и сохранение результата в DataFrame
books_count = execute_sql_query(query)
books_count


book_count
0         819

819 книг вышло после 1 января 2000 года

### Для каждой книги посчитаем количество обзоров и среднюю оценку

In [9]:
# SQL-запрос
query = """
    SELECT 
        b.book_id,
        b.title,
        COUNT(DISTINCT r.review_id) AS num_reviews,
        AVG(rt.rating) AS avg_rating
    FROM 
        books b
    LEFT JOIN 
        reviews r ON b.book_id = r.book_id
    LEFT JOIN 
        ratings rt ON b.book_id = rt.book_id
    GROUP BY 
        b.book_id,
        b.title
    ORDER BY num_reviews DESC, avg_rating DESC
"""

# Выполнение SQL-запроса и сохранение результата в DataFrame
book_stats = execute_sql_query(query)
book_stats

book_id                                              title  num_reviews  \
0        948                            Twilight (Twilight  #1)            7   
1        302  Harry Potter and the Prisoner of Azkaban (Harr...            6   
2        299  Harry Potter and the Chamber of Secrets (Harry...            6   
3        656                                     The Book Thief            6   
4        734                                   The Glass Castle            6   
..       ...                                                ...          ...   
995      191  Disney's Beauty and the Beast (A Little Golden...            0   
996      387                               Leonardo's Notebooks            0   
997      221                          Essential Tales and Poems            0   
998       83    Anne Rice's The Vampire Lestat: A Graphic Novel            0   
999      808                            The Natural Way to Draw            0   

     avg_rating  
0      3.662500  
1      4.414634  
2      4.287500  
3      4.264151  
4      4.206897  
..          ...  
995    4.000000  
996    4.000000  
997    4.000000  
998    3.666667  
999    3.000000  

[1000 rows x 4 columns]

Для каждой из тысячи книг определили количество обзоров и среднюю оценку.

In [10]:
# SQL-запрос
query = """
SELECT COUNT(*) AS count_review 
FROM reviews
"""
# Выполнение SQL-запроса и сохранение результата в DataFrame
count_review = execute_sql_query(query)
count_review

count_review
0          2793

2793 на 1000 книг было сделано

### Определим издательство, которое выпустило наибольшее число книг толще 50 страниц — так мы исключим из анализа брошюры

In [11]:
# SQL-запрос
query = '''
SELECT publishers.publisher, COUNT(books.book_id) as num_books
FROM books
JOIN publishers ON books.publisher_id = publishers.publisher_id
WHERE books.num_pages > 50
GROUP BY publishers.publisher
ORDER BY num_books DESC
LIMIT 1;
'''
# Выполнение SQL-запроса и сохранение результата в DataFrame
more_than_50 = execute_sql_query(query)
more_than_50

publisher  num_books
0  Penguin Books         42

Penguin Books - издательство, которое выпустило наибольшее число книг.

### Определим автора с самой высокой средней оценкой книг — учитываем только книги с 50 и более оценками

In [12]:
# SQL-запрос
query = '''
SELECT books.author_id, authors.author, AVG(ratings.rating) as avg_rating
FROM books
JOIN authors ON books.author_id = authors.author_id
JOIN ratings ON books.book_id = ratings.book_id
WHERE ratings.rating IS NOT NULL AND books.book_id IN (
  SELECT book_id
  FROM ratings
  GROUP BY book_id
  HAVING COUNT(rating_id) >= 50
)
GROUP BY books.author_id, authors.author
ORDER BY avg_rating DESC
LIMIT 1;
'''

# Выполнение SQL-запроса и сохранение результата в DataFrame
highest_rating = execute_sql_query(query)
highest_rating

author_id                      author  avg_rating
0        236  J.K. Rowling/Mary GrandPré    4.287097

Джоан Роулинг - автор с самым высоким средним рейтингом произведений

### Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [13]:
# SQL-запрос
query = '''
SELECT AVG(num_reviews) as avg_num_reviews
FROM (
    SELECT username, COUNT(*) as num_reviews
    FROM reviews
    WHERE username IN (
        SELECT username
        FROM ratings
        GROUP BY username
        HAVING COUNT(*) > 50
    )
    GROUP BY username
) as subquery;
'''

# Выполнение SQL-запроса и сохранение результата в DataFrame
mean_review_count = execute_sql_query(query)
mean_review_count

avg_num_reviews
0        24.333333

Пользователи, поставившие больше 50 оценок, в среднем написали по 24 обзора.

## Общий вывод

В ходе изучения баз данных мы проделали следующую работу:
    
- Определили, сколько книг вышло после 1 января 2000 года - 819.
- Определили количество обзоров и среднюю оценку для каждой из тысячи книг. Всего 2793 ревью.
- Определили, что Penguin Books - издательство, которое выпустило наибольшее число книг.
- Определили, что Джоан Роулинг - автор с самым высоким средним рейтингом прозведений
- Определили, что пользователи, поставившие больше 50 оценок, в среднем написали по 24 обзора.

## Рекомендации

**Основываясь на вышеуказанных фактах, я могу предложить следующие рекомендации:**

Разработать стратегию для продвижения и продажи книг, которые вышли после 1 января 2000 года. Так как их количество составляет 819, это может быть ценный сегмент рынка, на который стоит обратить внимание и разработать специальные маркетинговые активности.

Анализировать и улучшать обзоры и оценки пользователей для каждой из тысячи книг. Высокий уровень обзоров и положительные оценки могут способствовать росту продаж и популярности книг. Необходимо отслеживать и анализировать обратную связь от пользователей и предпринимать меры для улучшения качества обзоров и оценок.

Установить партнерство или сотрудничество с издательством Penguin Books, так как они выпустили наибольшее число книг. Это может предоставить возможности для совместных маркетинговых и продажных активностей, а также расширить ассортимент предлагаемых книг от этого издательства.

Уделить внимание автору Джоан Роулинг, так как она имеет самый высокий средний рейтинг произведений. Это может быть индикатором ее популярности среди пользователей и потенциально может привести к увеличению продаж ее книг. Разработать маркетинговые стратегии, связанные с произведениями этого автора, такие как кампании по продвижению новых книг или рекламные активности, связанные с автором.

Анализировать и взаимодействовать с пользователями, которые оставляют много оценок и пишут обзоры. Они могут быть ценными ресурсами для продвижения книг, так как в среднем они оставляют 24 обзора. Можно разработать программу лояльности или предложить бонусы таким пользователям, чтобы стимулировать их активность и вовлеченность в продвижении книг.

Дополнительно исследовать и анализировать другие данные из базы данных, такие как жанры книг, их продажи, взаимодействие с пользователями и другие факторы, которые могут влиять на успех книг. На основе этих данных можно выявить тенденции и предпочтения пользователей, определить наиболее популярные жанры книг, и использовать эту информацию для разработки стратегий продвижения и маркетинговых кампаний.

Улучшить взаимодействие с пользователями, особенно теми, кто оставляет обзоры и оценки. Это может включать улучшение пользовательского опыта на веб-сайте или в приложении, предоставление более удобных и простых способов оставлять обзоры и оценки, и активное взаимодействие с пользователями на социальных медиа и других платформах.

Разработать программу лояльности или бонусную систему для поощрения пользователей, оставляющих обзоры и оценки. Это может включать предоставление скидок, подарков или других привилегий для активных пользователей, что может стимулировать их участие и лояльность к бренду.

Использовать данные о пользователях, оставивших более 50 оценок, для создания сегментов аудитории и настройки таргетированных маркетинговых кампаний. Такие пользователи могут быть более заинтересованы в новых книгах, обзорах и рекомендациях, поэтому предоставление им персонализированного контента и предложений может увеличить их активность и вовлеченность.

Внимательно отслеживать и анализировать конкурентную среду, включая другие издательства и авторов, чтобы адаптироваться к изменениям на рынке и оставаться конкурентоспособными. Это может включать мониторинг новых релизов, маркетинговых активностей конкурентов и анализ отзывов и рейтингов их книг, чтобы определить свои преимущества и разработать соответствующие маркетинговые стратегии.